server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -697,10 +697,7 @@ private Snapshot newMultiSnapshot(TranslogSnapshot[] snapshots) throws IOExcepti
             "callers of readersAboveMinSeqNo must hold a lock: readLock ["
                 + readLock.isHeldByCurrentThread() + "], writeLock [" + readLock.isHeldByCurrentThread() + "]";
         return Stream.concat(readers.stream(), Stream.of(current))
-            .filter(reader -> {
-                final long maxSeqNo = reader.getCheckpoint().maxSeqNo;
-                return maxSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || maxSeqNo >= minSeqNo;
-            });
+            .filter(reader -> minSeqNo <= SequenceNumbers.min(reader.getCheckpoint().trimmedAboveSeqNo, reader.getCheckpoint().maxSeqNo));
```

Comment:
I think we need to account for the case where trimmedAboveSeqNo is equal to `SequenceNumbers.UNASSIGNED_SEQ_NO`. Perhaps we can add a method to checkpoint called `getMaxEffectiveSeqNo` which is:
```
if (trimmedAboveSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO)
  return maxSeqNo;
else
  return Math.min(maxSeqNo, trimmedAboveSeqNo);
```

Reply:
`SequenceNumbers.min` should handle this case:
https://github.com/elastic/elasticsearch/blob/044139a8432f9d14190de365402f0a0da1212b18/server/src/main/java/org/elasticsearch/index/seqno/SequenceNumbers.java#L84-L85

Having an explicit method is better. I pushed https://github.com/elastic/elasticsearch/pull/50205/commits/563545c5ee3629f5e2816463020815716aa65e17.

Reply:
Ah lol, I did not see that a special `min` method was used (I read it as Math.min).